src/helper/helper.js

```diff
@@ -69,13 +69,33 @@ class DataLayerHelper {
    * Creates a new helper object for the given dataLayer.
    *
    * @param {!Array<*>} dataLayer The dataLayer to help with.
-   * @param {function(!Object<*>, !Object<*>)=} listener The callback
-   *     function to execute when a new state gets pushed onto the dataLayer.
-   * @param {boolean=} listenToPast If true, the given listener will be
-   *     executed for state changes that have already happened.
+   * @param {{listener: function(!Object<*>, !Object<*>),
+   *     listenToPast: boolean,
+   *     processNow: boolean,
+   *     commandProcessors: !Object<string,
+   *         !Array<function(*):(!Object|undefined)>>,}=} optionsObject
    */
-  constructor(dataLayer, listener = () => {
-  }, listenToPast = false) {
+  constructor(dataLayer, {
+    // The callback function to execute when a new state
+    // gets pushed onto the dataLayer.
+    'listener': listener = (() => {}),
+    // If true, the given listener will be
+    // executed for state changes that have already happened.
+    'listenToPast': listenToPast = false,
+    // If false, halts the setup of the helper and delays processing the
+    // dataLayer until the 'process' method is called. Primarily used to
+    // register processors after construction of the helper.
+    'processNow': processNow = true,
+    'commandProcessors': commandProcessors = {},
+  } = {}) {
```

Comment:
IIUC, this is destructuring the options object into each option, without necessarily giving a name to the options itself.

I haven't run into it before, but this does seem allowed by the style guide: https://google.github.io/styleguide/jsguide.html#features-objects-destructuring

I still feel like in this case, when we expect arguments[1] to be either an options or a function, destructuring feels a bit weird.  For example, if someone passed a function with processNow hanging off it, what should happen?

I think given that I'd prefer a simple named object and examining it to determine if it is the new or old constructor format.

e.g.:
```
/**
 * @param {!DataLayer} dataLayer
 * @param {(!DataLayerOptions|function(): undefined)=} options
 * @param {boolean=} listenToPast
function (dataLayer, options = {}, listenToPast = undefined) {
  if (typeof options === 'function') {
    // Legacy invocation.
    options = {
      listener: options,
      listenToPast: listenToPast,
    };
  }
  
  this.listener_ = options.listener;
  // etc...
}
```

Reply:
I went ahead and updated it to be similar to your suggestion here, but I'm having some trouble with the default options. Ideally we'd want options that aren't explicitly passed in to default to the correct values (the most notable being processNow which needs to default to true). Unfortunately I can't find an elegant way to set these default values, especially since without string literals closure optimizes the intended option names away.

I came up with a bulky solution:
```
    if (typeof options === 'function') {
      options = {
        listener: options,
        listenToPast: listenToPast,
        processNow: true,
        commandProcessors: {},
      };
    } else {
      options = {
        listener: options['listener'] === undefined ?
            () => {} : options['listener'],
        listenToPast: options['listenToPast'] === undefined ?
            false : options['listenToPast'],
        processNow: options['processNow'] === undefined ?
            true : options['processNow'],
        commandProcessors: options['commandProcessors'] === undefined ?
            {} : options['commandProcessors'],
      };
    }
```
But I figured you might know of a way to improve upon it.

Reply:
The example you have is basically how the compiler rewrites the destructuring to work in ES5.  However, sometimes it's not necessary to do an explicit undefined check, like when all expected values are truthy.  For example, instead of
```js
listener: options['listener'] === undefined ? () => {} : options['listener'],
```

If the option is not set, it'll be falsey (undefined), and if it's set to a function, it'll be truthy.  So you can provide the default value using logical or:
```js
listener: options['listener'] || () => {},
```

This gets a bit tricky when options should default to true when undefined, like processNow, but something similar should still work:
```js
 // True when options['processNow'] is undefined or any truthy value.
processNow: options['processNow'] ===  undefined || !!options['processNow'], 
```

These are just shorthand though, only using the default when the option is explicitly undefined is also fine.